src/app/util/attribute-storage.cpp

```diff
@@ -261,9 +261,43 @@ uint16_t emberAfGetDynamicIndexFromEndpoint(EndpointId id)
     return kEmberInvalidEndpointIndex;
 }
 
+const EmberAfCluster * getClusterTypeDefinition(EndpointId endpointId, ClusterId clusterId, EmberAfClusterMask mask)
+{
+    uint16_t index = emberAfIndexFromEndpointIncludingDisabledEndpoints(endpointId);
+    if (index != kEmberInvalidEndpointIndex)
+    {
+        return emberAfFindClusterInType(emAfEndpoints[index].endpointType, clusterId, mask, nullptr);
+    }
+    // not found
+    return nullptr;
+}
+
+CHIP_ERROR setupDynamicEndpointDeclaration(EmberAfEndpointType & endpointType, EndpointId templateEndpointId,
+                                           const Span<const ClusterId> & templateClusterIds)
+{
+    // allocate cluster list
+    endpointType.clusterCount = static_cast<uint8_t>(templateClusterIds.size());
+    endpointType.cluster      = new EmberAfCluster[endpointType.clusterCount];
+    endpointType.endpointSize = 0;
+    // get the actual cluster pointers and sum up memory size
+    for (size_t i = 0; i < templateClusterIds.size(); i++)
+    {
+        auto cluster = getClusterTypeDefinition(templateEndpointId, templateClusterIds.data()[i], 0);
```

Comment:
Does just operator[] work?
```suggestion
        auto cluster = getClusterTypeDefinition(templateEndpointId, templateClusterIds.[i], 0);
```

Reply:
Needs a comment of what `0` is. If we always pass in the same value to the method call, maybe the parameter should not exist.

Reply:
- []: yes, corrected.
- the `mask` parameter is an option for subselecting clusters in the underlying `emberAfFindClusterInType`, so I kept that option in the new utility function, altough there's no immediate need for having it now.

Reply:
Please apply YAGNI: if you do not need it now, unless you have a pending PR that uses it right away, please do not add stuff for "will be needed later". Most of the time the "later" never comes.